### PR TITLE
Created three new routes on site router

### PR DESCRIPTION
**Pull request by florisdobber**
Ticket link: [https://code4community-team.monday.com/boards/925293934/pulses/1185013981](https://code4community-team.monday.com/boards/925293934/pulses/1185013981)

### CODE REVIEW

File: api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  public List<String> validateFields(String fieldPrefix) throws HandledException {
+    String fieldName = fieldPrefix + "record_stewardship_request.";
+    List<String> fields = new ArrayList<>();
+
+    if (date == null) {
+      fields.add(fieldName + "date");
+    }
+    if (!(watered || mulched || cleaned || weeded)) {
+      fields.add(fieldName + "activities");
+    }
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java:94]: Maybe also a `(duration == null || duration > 0)`

File: api/src/main/java/com/codeforcommunity/rest/subrouter/SiteRouter.java [hunk 2]
```diff
@@ -17,6 +24,53 @@ public SiteRouter(ISiteProcessor processor) {
...
+  }
+
+  private void handleRecordStewardshipRoute(RoutingContext ctx) {
+    JWTData userData = ctx.get("jwt_data");
+    int siteId = RestFunctions.getRequestParameterAsInt(ctx.request(), "site_id");
+
+    RecordStewardshipRequest recordStewardshipRequest =
+        RestFunctions.getJsonBodyAsClass(ctx, RecordStewardshipRequest.class);
+
+    processor.recordStewardship(userData, siteId, recordStewardshipRequest);
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/rest/subrouter/SiteRouter.java:74]: Need an end to this function

File: service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java [hunk 3]
```diff
@@ -10,4 +19,58 @@
...
+    record.setUserId(userData.getUserId());
+    record.setSiteId(siteId);
+    record.store();
+  }
+
+  @Override
+  public void unfavoriteSite(JWTData userData, int siteId) {
+    checkSiteExists(siteId);
+    if (isAlreadyFavorite(userData.getUserId(), siteId)) {
+      throw new ResourceDoesNotExistException(-1, "FavoriteSiteRecord");
```
Comment (liammoyn) [inline service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java:53]: Could replace this with a WrongFavoriteStatus type exception that can be used in the favorite site function too

File: service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java [hunk 4]
```diff
@@ -10,4 +19,58 @@
...
+    FavoriteSitesRecord record = db.newRecord(FAVORITE_SITES);
+    record.setUserId(userData.getUserId());
+    record.setSiteId(siteId);
+    record.store();
+  }
+
+  @Override
+  public void unfavoriteSite(JWTData userData, int siteId) {
+    checkSiteExists(siteId);
+    if (isAlreadyFavorite(userData.getUserId(), siteId)) {
```
Comment (liammoyn) [inline service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java:52]: Should be !isAlreadyFavorite

**Review by liammoyn — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java (hunk 1 above)
Comment (florisdobber) [reply api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java:94]: Duration is allowed to be null

File: api/src/main/java/com/codeforcommunity/rest/subrouter/SiteRouter.java (hunk 2 above)
Comment (florisdobber) [reply api/src/main/java/com/codeforcommunity/rest/subrouter/SiteRouter.java:74]: My bad

File: service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java (hunk 3 above)
Comment (florisdobber) [reply service/src/main/java/com/codeforcommunity/processor/SiteProcessorImpl.java:53]: Done

File: api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Override
+  public List<String> validateFields(String fieldPrefix) throws HandledException {
+    String fieldName = fieldPrefix + "record_stewardship_request.";
+    List<String> fields = new ArrayList<>();
+
+    if (date == null) {
+      fields.add(fieldName + "date");
+    }
+    if (!(watered || mulched || cleaned || weeded)) {
```
Comment (connernilsen) [inline api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java:92]: We have `Boolean` boxed values, so these can be null

**Review by connernilsen — COMMENTED** [review]
Looks good, just one comment about the DTO

File: api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java (hunk 5 above)
Comment (florisdobber) [reply api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java:92]: Switched to `boolean`

File: api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java (hunk 1 above)
Comment (liammoyn) [reply api/src/main/java/com/codeforcommunity/dto/site/RecordStewardshipRequest.java:94]: Yeah I got my conditional confused, I meant if duration is not null then it should be positive, so `(duration != null && duration < 0)`

File: api/src/main/java/com/codeforcommunity/rest/FailureHandler.java [hunk 6]
```diff
@@ -1,24 +1,6 @@
...
-import com.codeforcommunity.exceptions.MissingHeaderException;
-import com.codeforcommunity.exceptions.MissingParameterException;
-import com.codeforcommunity.exceptions.ResourceDoesNotExistException;
-import com.codeforcommunity.exceptions.TokenInvalidException;
-import com.codeforcommunity.exceptions.UsedSecretKeyException;
-import com.codeforcommunity.exceptions.UserDoesNotExistException;
-import com.codeforcommunity.exceptions.UserNotOnTeamException;
-import com.codeforcommunity.exceptions.UsernameAlreadyInUseException;
-import com.codeforcommunity.exceptions.WrongTeamRoleException;
+import com.codeforcommunity.exceptions.*;
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/rest/FailureHandler.java:3]: Aaahh no star

File: api/src/main/java/com/codeforcommunity/rest/FailureHandler.java [hunk 7]
```diff
@@ -210,6 +192,18 @@ public void handleMemberStatusException(RoutingContext ctx, MemberStatusExceptio
     end(ctx, message, 400);
   }
 
+  public void handleWrongFavoriteStatusException(RoutingContext ctx, WrongFavoriteStatusException e) {
+    String message = "";
```
Comment (liammoyn) [inline api/src/main/java/com/codeforcommunity/rest/FailureHandler.java:196]: You don't have to initialize this variable as an empty string, just leave it as a declaration: `String message;`

File: api/src/main/java/com/codeforcommunity/rest/FailureHandler.java (hunk 6 above)
Comment (florisdobber) [reply api/src/main/java/com/codeforcommunity/rest/FailureHandler.java:3]: I turned this off don't know how it snuck in!

File: api/src/main/java/com/codeforcommunity/rest/FailureHandler.java (hunk 7 above)
Comment (florisdobber) [reply api/src/main/java/com/codeforcommunity/rest/FailureHandler.java:196]: Didn't know that, thanks!

**Review by liammoyn — APPROVED** [review]

**Review by connernilsen — APPROVED** [review]
Looks good. Quick question though, have you double checked to see what happens when you omit or set the `boolean` values to null in a request, and does it have the response you want it to have?

**Comment by florisdobber** [comment]
> 
> 
> Looks good. Quick question though, have you double checked to see what happens when you omit or set the `boolean` values to null in a request, and does it have the response you want it to have?

Yes, checked and it does.

**Pull request merged**